R/slide.R

```diff
@@ -95,29 +105,42 @@
 #'  # slide a 7-day trailing average formula on cases
 #'   jhu_csse_daily_subset %>%
 #'   group_by(geo_value) %>%
-#'   epi_slide(cases_7dav = mean(cases), n = 7, 
-#'             align = "right") %>% 
+#'   epi_slide(cases_7dav = mean(cases), before = 6) %>% 
 #'   # rmv a nonessential var. to ensure new col is printed
 #'   dplyr::select(-death_rate_7d_av) 
 #'  
-#'  # slide a left-aligned 7-day average
+#'  # slide a 7-day leading average
 #'   jhu_csse_daily_subset %>%
 #'   group_by(geo_value) %>%
-#'   epi_slide(cases_7dav = mean(cases), n = 7, 
-#'             align = "left") %>% 
+#'   epi_slide(cases_7dav = mean(cases), before = 0, after = 6) %>% 
+#'   # rmv a nonessential var. to ensure new col is printed
+#'   dplyr::select(-death_rate_7d_av)
+#'   
+#'  # slide a 7-day centre-aligned average
+#'   jhu_csse_daily_subset %>%
+#'   group_by(geo_value) %>%
+#'   epi_slide(cases_7dav = mean(cases), before = 3, after = 3) %>% 
+#'   # rmv a nonessential var. to ensure new col is printed
+#'   dplyr::select(-death_rate_7d_av) 
+#'   
+#'  # slide a 14-day centre-aligned average
+#'   jhu_csse_daily_subset %>%
+#'   group_by(geo_value) %>%
+#'   epi_slide(cases_7dav = mean(cases), before = 6, after = 7) %>% 
 #'   # rmv a nonessential var. to ensure new col is printed
 #'   dplyr::select(-death_rate_7d_av) 
 #'  
 #'  # nested new columns
-#'  jhu_csse_daily_subset %>% 
-#'  group_by(geo_value) %>%
-#'  epi_slide(a = data.frame(cases_2dav = mean(cases), 
+#'   jhu_csse_daily_subset %>% 
+#'   group_by(geo_value) %>%
+#'   epi_slide(a = data.frame(cases_2dav = mean(cases), 
 #'                           cases_2dma = mad(cases)),
-#'            n = 2, as_list_col = TRUE)
-epi_slide = function(x, f, ..., n, ref_time_values,
-                     align = c("right", "center", "left"), before, time_step, 
+#'             before = 1, as_list_col = TRUE)
```

Comment:
I feel like `before` isn't special. I think better would make both = 0 by default. Then nothing is ever missing and we don't need the warning.

Reply:
I was bringing this up to Logan, who disagreed and claimed that a person wouldn't want an "identity" map when neither `before` nor `after` is specified, despite its not producing bugs.

Reply:
I can't seem to find where this was discussed. @brookslogan if this is your preference, feel free to approve.

Reply:
Issue for this is https://github.com/cmu-delphi/epiprocess/issues/161, but the discussion/decision is actually [here](https://github.com/cmu-delphi/epiprocess/issues/146#issuecomment-1192785302).  The idea was that trailing windows are the most common choice, so passing just `before` should be possible.  (Ken also made it possible to just pass `after`, so `before` actually isn't special; maybe we need to update the defaults that the user sees to make this clear.)  I think it's much more likely that passing no before&after is a bug than the user would want to have a 0-width window.

Reply:
Sorry, I forgot what we actually settled on; `before` is special, and passing `after` without `before` is the warning case.

Reply:
This is a little complex and asymmetric, but use cases are also asymmetric.
- before = k, after = 0: very common.  Used in version-unaware forecasting, forecast covariate preparation, growth rate calculations, direction calculations, maybe outlier detection, maybe week-level target definitions (Saturdays as representative dates).
- before = 0, after = k: uncommon or common.  Used in growth rate and direction calculations, maybe week-level target definitions (Sundays as representative date).
- before = k, after = k: common.  Used for more representative time values for calculations (e.g., maybe forecast covariates), outlier detection, maybe week-level target definitions (Wednesdays as representative dates).
- before = 0, after = 0: extremely uncommon.  Can't think of a use case.  If `epi_slide` was more like summarize, maybe it could be used for geo aggregation, but I don't think it could be used that way now.

I think an error/warning on missing before && missing after should stay.  Decisions to be made:
- Do we keep a warning on `before` missing, `after` nonzero?
  - Sketching out the above makes me think not.  Leading windows aren't all that rare relative to centered windows, and this warning makes the interface more asymmetric.
- How do we express the defaults in the function signature?  E.g., assuming we take out the warning above:
  - `before, after`: user might be confused by `before`, `after`, `ref_time_values`, and `time_step` with no defaults. --- I slightly favor this option.
  - `before=0, after=0`: then user might be surprised by the error when they don't pass either, because it looks like they are already taken care of from the function signature.  (Plus manually passing 0s for both to address this would be even more mysterious.)
  - `before, after=0`: doesn't make sense if we have taken out the warning above
- Do we error/warn on `before` missing, user-provided `after=0`?
  - I suggest a warning here.  A user is either (a) new and trying to get a trailing window, and forgetting to set `before` or expecting some default (--> error/warning), or (b) tweaking parameters of an existing trailing window to see if a 0-width "window" is actually better than another value that they were trying (--> warning/nothing).
- Do we error/warn on `after` missing, user-provided `before=0`?
  - I suggest a warning here.

Reply:
I will update this to take into consideration what you suggest.